sdk/src/main/java/com/mapbox/maps/renderer/widget/BitmapWidgetRenderer.kt

```diff
@@ -0,0 +1,326 @@
+
+      GLES20.glUniform1i(uniformTexture, 0)
+
+      GLES20.glClearColor(1f, 1f, 1f, 1f)
+
+      GLES20.glEnableVertexAttribArray(attributeVertexPosition)
+      GlUtils.checkError("glEnableVertexAttribArray")
+
+      GLES20.glVertexAttribPointer(
+        attributeVertexPosition, COORDS_PER_VERTEX,
+        GLES20.GL_FLOAT, false,
+        VERTEX_STRIDE, vertexPositionBuffer
+      )
+      GlUtils.checkError("glVertexAttribPointer")
+
+      GLES20.glEnableVertexAttribArray(attributeTexturePosition)
+      GlUtils.checkError("glEnableVertexAttribArray")
+
+      GLES20.glVertexAttribPointer(
+        attributeTexturePosition, COORDS_PER_VERTEX,
+        GLES20.GL_FLOAT, false,
+        VERTEX_STRIDE, texturePositionBuffer
+      )
+      GlUtils.checkError("glVertexAttribPointer")
+
+      GLES20.glDrawArrays(GLES20.GL_TRIANGLE_STRIP, 0, VERTEX_COUNT)
+      GlUtils.checkError("glDrawArrays")
+
+      GLES20.glDisableVertexAttribArray(attributeVertexPosition)
+      GLES20.glDisableVertexAttribArray(attributeTexturePosition)
+      GLES20.glBindBuffer(GLES20.GL_ARRAY_BUFFER, 0)
+      GLES20.glUseProgram(0)
+    }
+    needRender = false
+  }
+
+  override fun deinitialize() {
+    if (program != 0) {
+      GLES20.glDisableVertexAttribArray(attributeVertexPosition)
+      GLES20.glDetachShader(program, vertexShader)
+      GLES20.glDetachShader(program, fragmentShader)
+      GLES20.glDeleteShader(vertexShader)
+      GLES20.glDeleteShader(fragmentShader)
+      GLES20.glDeleteTextures(textures.size, textures, 0)
+      GLES20.glDeleteProgram(program)
+      program = 0
+    }
+    needRender = false
+  }
+
+  private fun textureFromBitmap() {
+    if (!bitmap.isRecycled) {
+      GLES20.glGenTextures(1, textures, 0)
+      GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, textures[0])
+      GLES20.glTexParameterf(
+        GLES20.GL_TEXTURE_2D,
+        GLES20.GL_TEXTURE_MIN_FILTER,
+        GLES20.GL_NEAREST.toFloat()
+      )
+      GLES20.glTexParameterf(
+        GLES20.GL_TEXTURE_2D,
+        GLES20.GL_TEXTURE_MAG_FILTER,
+        GLES20.GL_LINEAR.toFloat()
+      )
+      GLES20.glTexParameterf(
+        GLES20.GL_TEXTURE_2D,
+        GLES20.GL_TEXTURE_WRAP_S,
+        GLES20.GL_CLAMP_TO_EDGE.toFloat()
+      )
+      GLES20.glTexParameterf(
+        GLES20.GL_TEXTURE_2D,
+        GLES20.GL_TEXTURE_WRAP_T,
+        GLES20.GL_CLAMP_TO_EDGE.toFloat()
+      )
+      GLUtils.texImage2D(GLES20.GL_TEXTURE_2D, 0, bitmap, 0)
+      bitmap.recycle()
+    }
+  }
+
+  fun updateBitmap(bitmap: Bitmap) {
+    this.bitmap = bitmap
+    this.bitmapWidth = bitmap.width
+    this.bitmapHeight = bitmap.height
+    updateVertexBuffer()
+    updateMatrix = true
+    needRender = true
+  }
+
+  fun rotate(angleDegrees: Float) {
+    Matrix.setIdentityM(rotationMatrix, 0)
+    Matrix.setRotateM(rotationMatrix, 0, angleDegrees, 0f, 0f, 1f)
+    updateMatrix = true
+    needRender = true
+  }
+
+  fun translate(translateX: Float, translateY: Float) {
```

Comment:
shouldn't it be override?

Reply:
Haven't added those methods in the base interface, only rendering related. Do you think worth to add them?

Reply:
it's hard to navigate to me but I did see them in some other interface, guess would be nice to align it. If several class do translation / rotation - perhaps create a separate interface for that?

Reply:
Added methods to the WidgetRenderer interface